script/v2/manual-runs/scWETHv2Rebalance.s.sol

```diff
@@ -0,0 +1,392 @@
+    //         adaptersToInvest.push(compoundV3Adapter);
+    //         allocationPercents.push(COMPOUNDV3_ALLOCATION_PERCENT);
+    //         targetLtv[compoundV3Adapter] = COMPOUNDV3_TARGET_LTV;
+
+    //         totalAllocationPercent += COMPOUNDV3_ALLOCATION_PERCENT;
+    //     }
+
+    //     require(totalAllocationPercent == 1e18, "totalAllocationPercent != 100%");
+    // }
+
+    struct RebalanceDataParams {
+        IAdapter adapter;
+        uint256 amount;
+        uint256 flashLoanAmount;
+        bool isSupplyBorrow;
+    }
+
+    RebalanceDataParams[] rebalanceDataParams;
+
+    function _investAmount() internal view returns (uint256 investAmount) {
+        uint256 float = weth.balanceOf(address(vault));
+        uint256 minimumFloat = vault.minimumFloatAmount();
+        // investAmount == 0 just reinvests profits
+        investAmount = float > minimumFloat ? float - minimumFloat : 0;
+    }
+
+    function _updateRebalanceParams() internal {
+        uint256 investAmount = _investAmount();
+
+        IAdapter[3] memory allAdapters = [morphoAdapter, compoundV3Adapter, aaveV3Adapter];
+
+        IAdapter adapter;
+        uint256 allocationPercent;
+        uint256 amount;
+        uint256 flashLoanAmount;
+        bool isSupplyBorrow;
+        for (uint256 i; i < allAdapters.length; i++) {
+            adapter = allAdapters[i];
+            allocationPercent = adapterAllocationPercent[adapter];
+
+            if (allocationPercent != 0) {
+                amount = investAmount.mulWadDown(allocationPercent);
+                // first check if allocation Percent is greater than zero
+                (flashLoanAmount, isSupplyBorrow) = _calcRebalanceFlashLoanAmount(adapter, amount);
+
+                rebalanceDataParams.push(RebalanceDataParams(adapter, amount, flashLoanAmount, isSupplyBorrow));
+            } else {
+                uint256 ltv = getLtv(adapter);
+                // even if there is no allocation here, we still want to check if a disinvest is needed in this protocol
+                if (ltv > targetLtv[adapter] + disinvestThreshold) {
+                    (flashLoanAmount, isSupplyBorrow) = _calcRebalanceFlashLoanAmount(adapter, 0);
+
+                    rebalanceDataParams.push(RebalanceDataParams(adapter, 0, flashLoanAmount, false));
+                }
+            }
+        }
+    }
+
+    function _rebalance() internal {
+        _updateRebalanceParams();
+        (bytes[] memory callData, uint256 totalFlashLoanAmount) = _getRebalanceCallData();
+
+        if (totalFlashLoanAmount > minimumInvestAmount.mulWadDown(getLeverage())) {
+            console2.log("---- Running Rebalance ----");
+            vault.rebalance(_investAmount(), totalFlashLoanAmount, callData);
+        }
+    }
+
+    function _getRebalanceCallData() internal returns (bytes[] memory, uint256) {
+        uint256 stEthRateTolerance = 0.9995e18;
+        uint256 n = rebalanceDataParams.length;
```

Comment:
is there a way to not use this "n"?

Reply:
I have updated the code a little bit. Moving the calldata to storage. But I am not sure if there is a way to not use this n.

Reply:
i feel like this is adding more complexity overall

Reply:
I dont see a more simpler way to do it without the n.